Frontend for Python Transforms

romeovs opened the pull request:
Sets up the front end code and tests for Python transforms.

<img width="1474" height="906" alt="Screenshot 2025-09-18 at 23 36 21" src="https://github.com/user-attachments/assets/b55188bd-3a84-468a-a7d5-48370cc9a779" />

<img width="1474" height="906" alt="Screenshot 2025-09-18 at 23 36 39" src="https://github.com/user-attachments/assets/82a16ba2-0cfe-4017-b16d-e89e10145919" />

nmezzopera committed:
feat: add Python transform frontend pages

qnkhuat committed:
Integerate with FE

wotbrew committed:
Fix wrong source field name used for showing script in transform view

wotbrew committed:
Revert "Fix wrong source field name used for showing script in transform view"

This reverts commit c7dfb541f30ba1c01ee1c9f3472d458155a4e22b.

wotbrew committed:
Fix edit button not doing anything for python scripts

Closes QUE-2300

qnkhuat committed:
python transforms now properly being scheduled

qnkhuat committed:
FE select table when creating python transform

qnkhuat committed:
ensure tables are unique in tables selector

qnkhuat committed:
source alignments

crisptrutski committed:
Better default schema, better docstring

qnkhuat committed:
fixex test run

bronsa committed:
add cancel button

bronsa committed:
better endpoint name

bronsa committed:
add cancel button to test python endpoint

bronsa committed:
working button

wotbrew committed:
Logs are viewable inline during execution of python transforms

bronsa committed:
feat: settings page with s3 configs

bronsa committed:
feat: split settings

bronsa committed:
remove placeholders

crisptrutski committed:
Auto-update Python transform function signature when table aliases change

- Automatically updates def transform(...) signature when aliases are modified
- Updates docstring Args section with aligned descriptions and table names
- Adds reset button to restore default slugified aliases
- Validates alias input to only allow valid Python identifiers
- Handles multi-line signatures for many parameters (>80 chars)
- Shows visual distinction between custom and auto-generated aliases
- Fixes code editor layout to prevent overlap with Run button

bronsa committed:
eh, rename to appease the munging

crisptrutski committed:
Some minor fixes around docstring rewriting

crisptrutski committed:
solve more fe lints

crisptrutski committed:
fix type errors

crisptrutski committed:
more type check fixes - including some smelly optional override stuff

qnkhuat committed:
fix delete transform + table

bronsa committed:
fix: don't remove indentation before """

bronsa committed:
chore: allow empty schemas

crisptrutski committed:
Critical fix: docstring indentation

romeovs committed:
Add first-class PythonTransformSource type

# Conflicts:
#	enterprise/frontend/src/metabase-enterprise/transforms/components/QueryEditor/EditorBody/EditorBody.tsx

romeovs committed:
Lay python editor out in two columns

romeovs committed:
Make python editor resizable

romeovs committed:
Move parseCSV to utils

romeovs committed:
Split off results into separate component

romeovs committed:
Add empty and loading state components

romeovs committed:
Split up ExecutionResult component

romeovs committed:
Move python execution to separate hook

romeovs committed:
Remove color in editor gutter

romeovs committed:
Increase indentUnit for python

wotbrew committed:
Separate file / s3 lifetimes / writes / reads from http calls (#63294)

- python-runner no longer simulates a single http api that does
everything (incl s3 interaction)
- Separates file lifetimes (shared storage) from execution
- Easier to see how to avoid reading into strings (e.g. for piping into
tmp files, follow up)

Earlier on we had a service that managed much of the I/O, metabase now
coordinates it so it makes little sense pretending there is still
an overall http api.
# Conflicts:
#	enterprise/backend/src/metabase_enterprise/transforms/api.clj
#	enterprise/backend/src/metabase_enterprise/transforms/execute.clj
#	enterprise/backend/src/metabase_enterprise/transforms/python_runner.clj
#	enterprise/backend/test/metabase_enterprise/transforms/python_runner_test.clj

romeovs committed:
Allow extension to be passed to CodeEditor

romeovs committed:
Add pandas completions

romeovs committed:
Use DataGrid for displaying results

qnkhuat committed:
Python library APIs (#63378)

Co-authored-by: Romeo Van Snick <romeo@romeovansnick.be>
# Conflicts:
#	enterprise/backend/src/metabase_enterprise/api/routes.clj
#	enterprise/backend/src/metabase_enterprise/transforms/python_runner.clj
#	enterprise/backend/test/metabase_enterprise/transforms/python_runner_test.clj
#	enterprise/frontend/src/metabase-enterprise/api/tags.ts
#	resources/migrations/057_update_migrations.yaml

romeovs committed:
Use normal title case

romeovs committed:
Add syntax highlighting to Python script on Transform page

romeovs committed:
Handle json-line output

romeovs committed:
Remove unused props

romeovs committed:
Rename PythonEditor to PythonQueryEditor

romeovs committed:
Move script rewriter to utils

romeovs committed:
Flatten Python transform page

romeovs committed:
Rename shared to common

romeovs committed:
Remove dead code

romeovs committed:
Reduce specialization in Transform editor

romeovs committed:
Remove unused props

romeovs committed:
Move PythonQueryEditor to top-level components

romeovs committed:
Nest Python editor components in PythonTransformEditor

romeovs committed:
Restructure Python transform editor

romeovs committed:
Split up component more

romeovs committed:
Add helpers for python editor

romeovs committed:
Disable default schema

romeovs committed:
Add test ids

romeovs committed:
Fix import path

romeovs committed:
Reuse transform save modal

qnkhuat committed:
transforms python module (#63503)

# Conflicts:
#	enterprise/backend/src/metabase_enterprise/transforms/api.clj
#	enterprise/backend/src/metabase_enterprise/transforms/execute.clj
#	enterprise/backend/src/metabase_enterprise/transforms/jobs.clj
#	enterprise/backend/src/metabase_enterprise/transforms/util.clj
#	enterprise/backend/src/metabase_enterprise/transforms_python/python_runner.clj
#	enterprise/backend/test/metabase_enterprise/transforms_python/api_test.clj
#	enterprise/backend/test/metabase_enterprise/transforms_python/e2e_test.clj
#	enterprise/backend/test/metabase_enterprise/transforms_python/python_runner_test.clj

romeovs committed:
Fix parameter

romeovs committed:
Add e2e tests for python common lib

romeovs committed:
Move python transform api to general transforms folder

romeovs committed:
Move type to metabase-api/types

romeovs committed:
Disable debugger for python transforms for now, allow to open it with query param

romeovs committed:
Disable debugger hotkey for python transforms

romeovs committed:
Remove cancelation code for now

# Conflicts:
#	enterprise/frontend/src/metabase-enterprise/transforms/components/RunButton/RunButton.tsx
#	enterprise/frontend/src/metabase-enterprise/transforms/pages/TransformPage/RunSection/RunSection.tsx

romeovs committed:
Add more generic RunInfo

romeovs committed:
Remove unused LogOutput

romeovs committed:
Set up python runner in CI (#63566)

* Check out with submodules in CI

* Include python-runner in docker compose for tests

* Set up python runner in CI

* Fix import

* Update python-runner
# Conflicts:
#	python-runner

romeovs committed:
Fix font sizes

romeovs committed:
Fix table layout

romeovs committed:
Revert "Remove unused LogOutput"

This reverts commit 90c07043449ca20fdfc126aec6e0feb5f8db4c64.

romeovs committed:
Add LogOutput back to runsection

romeovs committed:
Remove duplicate schema field

romeovs committed:
Use css instead of styled components

romeovs committed:
Fix LogOutput styling

bronsa committed:
cherry pick driver code (#63628)

* cherry pick driver code

* update changelog

* add type->database-type unit tests

* kondo

* create/insert table

* insert-from-jsonl test

* missing rename-tables for snowflake

* extend rename-tables-test to more drivers

* line breaks

* add driver python transform feature

* handle review comments

* kondo

* feat: replace string->val with a more generic insert-col->val

* remove non transactional impls

* chore: docsstrings

* TODO

* extra test

* fix extra test

* test all base-types

* fix sqls

romeovs committed:
Add transforms/python feature

romeovs committed:
Do not allow picking unsupported databases

romeovs committed:
Use slugify from metabase/lib/formatting

romeovs committed:
Add type for transform alias map

romeovs committed:
Fix some type imports

romeovs committed:
Refactor PythonDataPicker

# Conflicts:
#	enterprise/frontend/src/metabase-enterprise/transforms/components/PythonTransformEditor/PythonDataPicker/PythonDataPicker.tsx

romeovs committed:
Further refactors

romeovs committed:
Use data selector for database picker

romeovs committed:
Use data picker to pick tables

nemanjaglumac committed:
Add analytics to history revision and data reference (#63587)

qnkhuat committed:
setup ci properly (#63693)

romeovs committed:
Improve tests for python transform

romeovs committed:
Transform run cancellation (#62387)

* Add cancellation to transform api

* Use effect to set polling state

* Add cancellation button to run button

* Set up cancellation on transform page

* Use confirmation modal to cancel transform

* Add run cancel button to run list

* Make api endpoint take transform id instead of run id.

* Fix cancelation api

* Add canceled state to TransformRunStatus

* Add canceled state to transform page

* Add polling to RunListPage

* Remove comment in code

* Set status to canceling when you click cancel

* Add canceling status

* Poll when the transform is canceling

* Add copy for runs that are being canceled

* Manually track fetching status for running and canceling

* Handle canceling status in utils

* Avoid wrapping cancel button in run list

* Add test for canceling run on transform page

* Add e2e for canceling test from runs page

* Fix prop type

* Change copy to Canceling

* Add custom color to canceled state

* Remove mock code

* Increase cancelation tiemout

* Fix copy in text

* Fix snowplow event

* Remove void specifier and close modal in handler directyle

* Add optimistic update for run cancelation

* Avoid manually tracking cancelation status

* Avoid useEffect when changing isPolling

* Add in missing helper

* Remove another useEffect

* Add placeholder run when starting

---------

Co-authored-by: Eric Normand <ericwnormand@gmail.com>
Co-authored-by: Alexander Polyankin <alexander.polyankin@metabase.com>

romeovs committed:
call api proxy instead of internal call, also have settings page for transform

romeovs committed:
multiple tables

romeovs committed:
run test

romeovs committed:
dead code

romeovs committed:
Remove unused var

romeovs committed:
Fix duplicate imports

romeovs committed:
Use DatabaseId in python transform

romeovs committed:
Remove source/draft ambiguity in types

romeovs committed:
Move library types to metabase-types/api

romeovs committed:
Remove api transforms

romeovs committed:
Remove vibe-comment

romeovs committed:
Remove unused component

romeovs committed:
Validate source more

romeovs commented on enterprise/frontend/src/metabase-enterprise/api/transform-python.ts:
```diff
@@ -0,0 +1,34 @@
+
+import { EnterpriseApi } from "./api";
+
+export const pythonRunnerApi = EnterpriseApi.injectEndpoints({
```
These api's are temporary and will be removed.

romeovs committed:
Move TableInput to separate file

romeovs committed:
Remove TableOption type

chodorowicz committed:
Fix command z not working in connection string input (#63689)

- Stop preventing default event handling onPaste.
- Improve Cypress paste emulation to be compatible with React way of handling paste events.

romeovs committed:
Simplify props

romeovs committed:
Move AliasInput to separate file

romeovs committed:
Remove unused props

romeovs committed:
Add border to empty results

bronsa committed:
Transforms prepare for python (#63698)

* pull out transform run utils to prepare for python transforms

* fix: check running-run on transform-id, not run-id

* chore: put on a new line

bpander committed:
Fix #63629: Filter on cards causes an unnecessary horizontal scroll bar for public/static embedding (#63650)

romeovs committed:
Fix import

romeovs committed:
Remove extension

romeovs committed:
Move draft back to new python transform page

lorem--ipsum committed:
Closes EMB-50 tabs dimmed out on iOS (#63690)

romeovs committed:
Fix new python transform page nesting

crisptrutski committed:
S3 transport layer for python-runner (#63697)

crisptrutski committed:
Add s3 dependencies (#63706)

bronsa committed:
port prometheus/instrumentation (#63631)

* port prometheus metrics

* port instrumentation

* instrument existing transform paths

* wider buckets

* remove nil handling

* warn on nil bytes/rows

* add stage-type mapping

* typo

romeovs committed:
Simpify new transform page structure

romeovs commented on frontend/src/metabase/common/components/CodeEditor/utils.ts:
```diff
@@ -28,7 +37,7 @@ export function getLanguageExtension(language: CodeLanguage): Extension {
       return json();
     case "python":
-      return python();
+      return [python(), indentUnit.of("    ")];
```
This adds 4-space indents for Python

romeovs commented on enterprise/frontend/src/metabase-enterprise/transforms/components/PythonTransformEditor/PythonDataPicker/PythonDataPicker.tsx:
```diff
@@ -0,0 +1,285 @@
+          {t`Select the database that contains your source data`}
+        </Text>
+
+        <DatabaseDataSelector
```
This picker is just temporary until we have multi-db Python transforms, it will go away eventually.

romeovs committed:
Implement data picker according to designs

romeovs committed:
Further adjust to design file

romeovs committed:
Fix some e2e tests

romeovs committed:
Merge remote-tracking branch 'origin/proj-python/feature-branch' into proj-python/frontend

romeovs committed:
Clone python runner when running CI

romeovs committed:
Avoid starting the container in e2e tests twice

romeovs committed:
Hide python runner admin route when user does not have the feature for it

romeovs committed:
Remove unused variables

romeovs committed:
Move Python runner amdin page to transforms plugin

romeovs committed:
Do not show Python library link when the user does not have python transforms support

romeovs committed:
Hide Python transforms features when the user does not have transforms-python token

romeovs committed:
Use useHasTokenFeature hook where possible

romeovs committed:
Add transforms-python feature

romeovs committed:
Fix e2e tests

romeovs committed:
Do not include python-runner in test docker-compose yet

romeovs committed:
Merge branch 'proj-python/feature-branch' into proj-python/frontend

# Conflicts:
#	.github/scripts/setup-python-runner.sh

romeovs committed:
Tag tests for python transforms with @transforms-python

romeovs committed:
Add helper to update Python library

romeovs committed:
Add check for python output

romeovs committed:
Fix Revert on python library when it does not exist

romeovs committed:
Fix e2e test helper

romeovs committed:
Add python runner settings in CI

ranquild reviewed (commented):
Since it's a separate feature, wdyt about adding a separate `transforms-python` EE feature folder and moving most code there? And, optioanlly, you could rely on our plugin system so that you wouldn't need to check if the feature is enabled.

romeovs committed:
WIP

romeovs committed:
Translate pandas definitions

romeovs committed:
Add AnsiLogs component

romeovs committed:
Use AnsiLogs in Logs components

romeovs committed:
Use AnsiLogs in LogOutput component

romeovs committed:
Remove token feature gate in enterprise folder

romeovs committed:
Return alias instead of throwing

romeovs committed:
Check type in isPythonTransform assertion

romeovs committed:
Use isResourceNotFoundError

romeovs committed:
Make transforms-python it's own plugin

romeovs committed:
Merge branch 'proj-python/feature-branch' into proj-python/frontend

wotbrew committed:
Move transform ordering tests to the python module (#63827)

* Add transform ordering tests to the python module

- Consistent with api tests for python transforms being in the python
module
- In practice transforms module 'knows' about python, so this is
organizational only
- Copied test helpers for now, not sure if functions strong enough to be
shared

* Check :transforms/python

romeovs committed:
Add test for editing Python transform

romeovs committed:
Verify that running the transform with the library works

romeovs committed:
Add missing transforms-python feature to mocks

romeovs committed:
Simplify source layout

bronsa committed:
Proj python/table name len (#63842)

* feat: ensure temp-table-name respects max length

* add unit test

* add limit for missing drivers

* depend on it

* fix: use db instead of db-id

qnkhuat committed:
/s/localhost/localstack

romeovs committed:
Change setting type of python runner token

romeovs committed:
Fix tests

romeovs committed:
Use ConcreteTableIds in python transforms only

romeovs committed:
Deduplicate codemirror helpers

bronsa committed:
Proj ptyhon/table swap (#63878)

* add rename feature, rename-table! multimethod

* add table swap with atomic or rename or fallback

* changelog

* drop requirement

* ensure we're testing replace flow

* wait for completion, table already exists

* wip

* add rename-table-test

* fix for snowflake

romeovs committed:
Ignore errors in non-parseable JSON lines

romeovs committed:
Use useLocation hook to enable debugger

romeovs committed:
Use useErrorMessage helper

romeovs committed:
Move SourceSection to transforms-python plugin

romeovs committed:
Fix codemirror helpers

romeovs committed:
Remove undefined color

romeovs committed:
Fix types

romeovs committed:
Fix codemirror helpers more

romeovs committed:
Extract sourceDatabaseId helper

romeovs committed:
Fix e2e helpers

romeovs committed:
Fix error in table picker

romeovs committed:
Add missing test-id

romeovs committed:
Fix transform tests

romeovs committed:
Merge branch 'proj-python/feature-branch' into proj-python/frontend

# Conflicts:
#	e2e/support/helpers/e2e-custom-column-helpers.ts
#	enterprise/frontend/src/metabase-enterprise/transforms/pages/TransformPageLayout/TransformPageLayout.tsx
#	frontend/src/metabase-types/api/transform.ts

romeovs committed:
Validate python source

romeovs committed:
Fix issues after merge

romeovs committed:
Fix tag test

romeovs committed:
Fix other tag test

nemanjaglumac commented on enterprise/frontend/src/metabase-enterprise/transforms-python/pages/PythonLibraryEditorPage/PythonLibraryEditorPage.tsx:
```diff
@@ -0,0 +1,140 @@
+import { useEffect, useState } from "react";
+import { t } from "ttag";
+
+import Button from "metabase/common/components/Button";
```
@romeovs @ranquild Any particular reason we had to use the deprecated `Button` here instead of the Mantine one?
